fido/pronom/soap.py

```diff
@@ -0,0 +1,101 @@
+TNA_DOMAIN = 'nationalarchives.gov.uk'
+PRONOM_HOST = 'www.{}'.format(TNA_DOMAIN)
+PRONOM_NS = 'http://pronom.{}'.format(TNA_DOMAIN)
+SIG_NS = 'http://{}/pronom/SignatureFile'.format(PRONOM_HOST)
+
+NS = {
+    'soap': 'http://schemas.xmlsoap.org/soap/envelope/',
+    'xsi': 'http://www.w3.org/2001/XMLSchema-instance',
+    'xsd': 'http://www.w3.org/2001/XMLSchema',
+    'pronom': PRONOM_NS,
+    'sig': SIG_NS
+}
+
+HEADERS = {
+    'Host': PRONOM_HOST,
+    'User-Agent': 'PRONOM UTILS v{0} (OPF)'.format(__version__),
+    'Content-type': 'text/xml; charset="UTF-8"'
+}
+
+def get_pronom_sig_version():
+    """
+    Get PRONOM signature version.
+
+    Return latest signature file version number as an int.
+    Raises an HTTPError if there are problems.
+    """
+    try:
+        tree = _get_soap_ele_tree('getSignatureFileVersionV1')
+        ver_ele = tree.find('.//pronom:Version/pronom:Version', NS)
+        return int(ver_ele.text)
+    except Exception as e:
+        sys.stderr.write('get_pronom_sig_version(): unknown error: {}'.format(str(e)))
+        raise e
+
+def get_pronom_signature():
+    """
+    Get PRONOM signature.
+
+    Return a tuple comprising the latest signature XML file as string and a count
+    of the FileFormat elements contained as an integer.
+    Upon error, write to `stderr` and return the tuple [], False.
+    """
+    try:
+        tree = _get_soap_ele_tree('getSignatureFileV1')
+        for prefix, uri in NS.items():
+            ET.register_namespace(prefix, uri)
+        sigfile_ele = tree.find('.//pronom:SignatureFile', NS)
+        format_ele_len = len(sigfile_ele.findall('.//sig:FileFormat', NS))
+        if format_ele_len < 1:
+            sys.stderr.write("get_pronom_signature(): could not parse XML from SOAP response: file")
+            return [], False
+        return '<?xml version="1.0" encoding="UTF-8"?>\n' + ET.tostring(sigfile_ele, encoding='unicode'), format_ele_len
+    except Exception as e:
+        sys.stderr.write("get_pronom_signature(): unknown error: " + str(e))
+        raise e
+
+def _get_soap_ele_tree(soap_action):
+    soap_string = '{}<soap:Envelope xmlns:xsi="{}" xmlns:xsd="{}" xmlns:soap="{}"><soap:Body><{} xmlns="{}" /></soap:Body></soap:Envelope>'.format(XML_PROC, NS.get('xsi'), NS.get('xsd'), NS.get('soap'), soap_action, PRONOM_NS).encode(ENCODING)
+    soap_action = '\"{}:{}In\"'.format(PRONOM_NS, soap_action)
+    xml = _get_soap_response(soap_action, soap_string)
+    for prefix, uri in NS.items():
+        ET.register_namespace(prefix, uri)
+    return ET.fromstring(xml)
+
+def _get_soap_response(soap_action, soap_string):
+    req = urllib.request.Request('http://{}/pronom/service.asmx'.format(PRONOM_HOST), data=soap_string)
+    for key, value in HEADERS.items():
+        req.add_header(key, value)
+    req.add_header('Content-length', '%d' % len(soap_string))
+    req.add_header('SOAPAction', soap_action)
+    with urllib.request.urlopen(req) as response:
```

Comment:
If we've decided to abandon Python 2.7 support then ignore this, but urlopen couldn't be used as a context manager unless we wrap the call with `contextlib.closing`.

Reply:
See the 2.7 comment above, again am happy to revert.